rabbit-escape-render/src/rabbitescape/render/WaterRegionRenderer.java

```diff
@@ -0,0 +1,364 @@
+import rabbitescape.engine.util.CellDebugPrint;
+import rabbitescape.engine.util.LookupItem2D;
+import rabbitescape.engine.util.Position;
+import rabbitescape.render.gameloop.WaterDynamics;
+
+public class WaterRegionRenderer implements LookupItem2D
+{
+
+    public static final int contentsPerParticle = 50;
+    public static final int maxParticleCountChange = 2;
+    private static final int maxHeightChange = 1;
+
+    public static enum BounceSurface
+    {
+        VERTICAL,
+        HORIZONTAL,
+        LEFT_RISE,
+        RIGHT_RISE
+    }
+
+    public WaterRegion region;
+
+    private int targetWaterHeight = 0;
+    private int height = 0;
+    private int lastHeight = 0;
+
+    public boolean drawnLT, drawnLB, drawnR;
+
+    private final World world;
+    private final WaterDynamics dynamics;
+
+    public WaterRegionRenderer(WaterRegion region, World world, WaterDynamics dynamics )
+    {
+        this.region = region;
+        this.world = world;
+        this.dynamics = dynamics;
+    }
+
+    public void setWaterRegion( WaterRegion region )
+    {
+        this.region = region;
+    }
+
+    public WaterRegionRenderer adjacentRenderer( CellularDirection d )
+    {
+        return dynamics.lookupRenderer.getItemAt( region.x + d.xOffset, region.y + d.yOffset );
+    }
+
+    public boolean adjacentNull( CellularDirection d )
+    {
+        return null == adjacentRenderer( d );
+    }
+
+    public boolean adjacentWaterIsFalling( CellularDirection d )
+    {
+        WaterRegionRenderer adj = adjacentRenderer( d );
+        if ( null == adj )
+        {
+            return false;
+        }
+        return adj.isFalling();
+    }
+
+    /**
+     * Called once per game step
+     */
+    public void setTargetWaterHeight()
+    {
+        if ( isFalling() )
+        {
+            targetWaterHeight = 0;
+            return;
+        }
+        Block block = world.getBlockAt( region.x, region.y );
+        if ( null == block)
+        {
+            targetWaterHeight = region.getContents() / 32;
+        }
+        else if ( BehaviourTools.isSlope( block ) )
+        {
+            targetWaterHeight = triangleHeight( region.getContents() );
+        }
+        else
+        {
+            throw new RuntimeException( "Unexpected block type" );
+        }
+        targetWaterHeight = targetWaterHeight > 32 ? 32 : targetWaterHeight;
+    }
+
+    /**
+     * Called once per animation step
+     */
+    public void setWaterHeight()
+    {
+        if ( lastHeight + maxHeightChange < targetWaterHeight )
+        {
+            height = lastHeight + maxHeightChange;
+            lastHeight = height;
+            return;
+        }
+        if ( lastHeight - maxHeightChange > targetWaterHeight )
+        {
+            height = lastHeight - maxHeightChange;
+            lastHeight = height;
+            return;
+        }
+        height = targetWaterHeight;
+        lastHeight = height;
+    }
+
```

Comment:
By the way, your impl and mine above both end up with height == lastHeight at the end of the method.  Is that intended?